sympy/polys/domains/quaternion.py

```diff
@@ -0,0 +1,524 @@
+        m22 = (-sqb - sqc + sqd + sqa) * invs
+
+        tmp1 = q.b * q.c
+        tmp2 = q.d * q.a
+        m10 = 2 * (tmp1 + tmp2) * invs
+        m01 = 2 * (tmp1 - tmp2) * invs
+
+        tmp1 = q.b * q.d
+        tmp2 = q.c * q.a
+        m20 = 2 * (tmp1 - tmp2) * invs
+        m02 = 2 * (tmp1 + tmp2) * invs
+
+        tmp1 = q.c * q.d
+        tmp2 = q.b * q.a
+        m21 = 2 * (tmp1 + tmp2) * invs
+        m12 = 2 * (tmp1 - tmp2) * invs
+
+        return Matrix([[m00, m01, m02], [m10, m11, m12], [m20, m21,
+                                                          m22]])
+
+    def to_matrix_4x4(self, v):
+        """ Generates a 4x4 transformation matrix (used for rotation about a point
+        other than the origin) from a quaternion and the position vector of the point.
+
+        Example
+        ========
+
+        >>> from sympy.polys.domains.quaternion import Quaternion
+        >>> q = Quaternion(1, 2, 3, 4)
+        >>> print(q.to_matrix_4x4((1, 4, 8)))
+        Matrix([[-2/5, 2/5, 11/5, -89/5], [2, -1/5, 2, -66/5], [1, 14/5, 0, -21/5], [0, 0, 0, 1]])
+        """
+
+        q = self
+        sqa = sqrt(q.a ** 2)
+        sqb = sqrt(q.b ** 2)
+        sqc = sqrt(q.c ** 2)
+        sqd = sqrt(q.d ** 2)
+
+        invs = 1 / (sqa + sqb + sqc + sqd)
+        m00 = (sqb - sqc - sqd + sqa) * invs
+        m11 = (-sqb + sqc - sqd + sqa) * invs
+        m22 = (-sqb - sqc + sqd + sqa) * invs
+
+        tmp1 = q.b * q.c
+        tmp2 = q.d * q.a
+        m10 = 2 * (tmp1 + tmp2) * invs
+        m01 = 2 * (tmp1 - tmp2) * invs
+
+        tmp1 = q.b * q.d
+        tmp2 = q.c * q.a
+        m20 = 2 * (tmp1 - tmp2) * invs
+        m02 = 2 * (tmp1 + tmp2) * invs
+
+        tmp1 = q.c * q.d
+        tmp2 = q.b * q.a
+        m21 = 2 * (tmp1 + tmp2) * invs
+        m12 = 2 * (tmp1 - tmp2) * invs
+
+        (x, y, z) = v
+
+        m03 = x - x * m00 - y * m01 - z * m02
+        m13 = y - x * m10 - y * m11 - z * m12
+        m23 = z - x * m20 - y * m21 - z * m22
+        m30 = m31 = m32 = 0
+        m33 = 1
+
+        return Matrix([[m00, m01, m02, m03], [m10, m11, m12, m13],
+                       [m20, m21, m22, m23], [m30, m31, m32, m33]])
+
+class CustomLatexPrinter(LatexPrinter):
+
+    def _print_Quaternion(self, expr):
+        i, j, k = symbols('i j k')
+        return latex(expr.a) + '+' + latex(expr.b * i) + '+' + \
+            latex(expr.c * j) + '+' + latex(expr.d * k)
+
+class CustomStrPrinter(StrPrinter):
+    def _print_Quaternion(self, expr):
+        m = max(len(str(expr.a)), len(str(expr.b)),
+                len(str(expr.c)), len(str(expr.d)))
+        if(m < 20):
+            return str(expr.a) + ' + ' + '(' + str(expr.b) + ')i + ' \
+                + '(' + str(expr.c) + ')j + ' + '(' + str(expr.d) + ')k'
+
+        return str(expr.a) + '\n+ ' + '(' + str(expr.b) + ')i\n+ ' \
+            + '(' + str(expr.c) + ')j\n+ ' + '(' + str(expr.d) + ')k'
+
+Basic.__str__ = lambda self: CustomStrPrinter().doprint(self)
```

Comment:
Still is a bad idea

Reply:
Thanks for all the suggestions. I shall inherit from Expr and rename inverse to inv. I really need cos(x)^2 + sin(x)^2 to simply to 1 though. Do you know of a way I can implement a form of trigsimp that checks only for this?

As for the printing, I knew all along that it was a bad idea but I wasn't sure what to do.
I guess it is clear to me now.

I shall remove the custom printers and add the _print_Quaternion methods to StrPrinter , LatexPrinter etc.

Reply:
Can u tell me where exactly I should add tests. Should I add them in sympy.core.tests/test_args or in sympy.polys.agca/tests or in both?

Reply:
This class should not be located in `sympy/polys`, maybe in `sympy/physics`? or in `sympy/geometry`?